include/RAJA/util/zip_tuple.hpp

```diff
@@ -0,0 +1,348 @@
+  RAJA_HOST_DEVICE inline auto operator()(Iter&& iter) const
+    -> decltype(std::forward<Iter>(iter) += rhs)
+  {
+    return std::forward<Iter>(iter) += rhs;
+  }
+};
+
+template < typename difference_type >
+struct MinusEq
+{
+  const difference_type& rhs;
+  template< typename Iter >
+  RAJA_HOST_DEVICE inline auto operator()(Iter&& iter) const
+    -> decltype(std::forward<Iter>(iter) -= rhs)
+  {
+    return std::forward<Iter>(iter) -= rhs;
+  }
+};
+
+struct DeRef
+{
+  template< typename Iter >
+  RAJA_HOST_DEVICE inline auto operator()(Iter&& iter) const
+    -> decltype(*std::forward<Iter>(iter))
+  {
+    return *std::forward<Iter>(iter);
+  }
+};
+
+struct Swap
+{
+  template< typename T0, typename T1 >
+  RAJA_HOST_DEVICE inline int operator()(T0&& t0, T1&& t1) const
+  {
+    using camp::safe_swap;
+    safe_swap(std::forward<T0>(t0), std::forward<T1>(t1));
+    return 1;
+  }
+};
+
+struct IterSwap
+{
+  template< typename T0, typename T1 >
+  RAJA_HOST_DEVICE inline int operator()(T0&& t0, T1&& t1) const
+  {
+    using RAJA::safe_iter_swap;
+    safe_iter_swap(std::forward<T0>(t0), std::forward<T1>(t1));
+    return 1;
+  }
+};
+
+
+/*!
+    \brief Call f on each member of t (f(t)...).
+*/
+template < typename Tuple, typename F, camp::idx_t... Is >
+RAJA_HOST_DEVICE inline
+void zip_for_each_impl(Tuple&& t, F&& f, camp::idx_seq<Is...>)
+{
+  camp::sink(std::forward<F>(f)(std::forward<Tuple>(t).template get<Is>())...);
+}
+
+/*!
+    \brief Call f on each member of t0 and t1 (f(t0, t1)...).
+*/
+template < typename Tuple0, typename Tuple1, typename F, camp::idx_t... Is >
+RAJA_HOST_DEVICE inline
+void zip_for_each_impl(Tuple0&& t0, Tuple1&& t1, F&& f, camp::idx_seq<Is...>)
+{
+  camp::sink(std::forward<F>(f)(std::forward<Tuple0>(t0).template get<Is>(), std::forward<Tuple1>(t1).template get<Is>())...);
+}
+
+/*!
+    \brief Call f on each member of t (f(t)...).
+*/
+template < typename Tuple, typename F >
+RAJA_HOST_DEVICE inline
+void zip_for_each(Tuple&& t, F&& f)
+{
+  zip_for_each_impl(std::forward<Tuple>(t), std::forward<F>(f), typename camp::decay<Tuple>::IdxSeq{});
+}
+
+/*!
+    \brief Call f on each member of t0 and t1 (f(t0, t1)...).
+*/
+template < typename Tuple0, typename Tuple1, typename F >
+RAJA_HOST_DEVICE inline
+void zip_for_each(Tuple0&& t0, Tuple1&& t1, F&& f)
+{
+  static_assert(std::is_same<typename camp::decay<Tuple0>::IdxSeq, typename camp::decay<Tuple1>::IdxSeq>::value,
+      "Tuple0 and Tuple1 must have the same size");
+  zip_for_each_impl(std::forward<Tuple0>(t0), std::forward<Tuple1>(t1), std::forward<F>(f), typename camp::decay<Tuple0>::IdxSeq{});
+}
+
+
+/*!
+    \brief Tuple used by ZipIterator for storing multiple references and values.
+    Acts like a reference to its members allowing copy/move construction/assignment
+    based on the reference type of the zip_tuple.
+*/
+template < bool is_val, typename ... Ts >
+struct zip_tuple
+{
+  using value_type = RAJA::tuple<Ts...>;
```

Comment:
```cpp
concepts::enable_if<is_trivially_copyable<typename decay<T>::type>...>
```

Reply:
std::is_trivially_copyable is not supported by gcc <5, which is extra painful because other compilers that use such a gcc install also don't support it.

Reply:
Well that's just cute. 🙃

I did find the following source code of LLVM to be relevant -- specifically how they test for `is_trivially_copyable` here: https://llvm.org/doxygen/type__traits_8h_source.html

@trws what are your thoughts on adapting what LLVM does? It looks like we need basic compiler support for all of the `is_*able` to make the `is_trivially_*able` traits

Reply:
I think that makes good sense, we already require at least 4.9, so we should be good on that score I think.  It's unfortunate that we need it, but I don't see a reason to keep it out, it's looking more and more like we're going to want a complete set of non-builtin typetraits at some point. =/